consensus/qbft/core/roundchange.go

```diff
@@ -0,0 +1,291 @@
+			logger.Info("QBFT: Justification of ROUND-CHANGE messages failed")
+			return nil
+		}
+
+		log.Info("QBFT: handleRoundChange - broadcasting pre-prepare")
+		r := &Request{
+			Proposal:        proposal,
+			RCMessages:      roundChangeMessages,
+			PrepareMessages: prepareMessages,
+		}
+		c.sendPreprepareMsg(r)
+	}
+	return nil
+}
+
+// highestPrepared returns the highest Prepared Round and the corresponding Prepared Block
+func (c *core) highestPrepared(round *big.Int) (*big.Int, istanbul.Proposal) {
+	return c.roundChangeSet.highestPreparedRound[round.Uint64()], c.roundChangeSet.highestPreparedBlock[round.Uint64()]
+}
+
+// ----------------------------------------------------------------------------
+
+func newRoundChangeSet(valSet istanbul.ValidatorSet) *roundChangeSet {
+	return &roundChangeSet{
+		validatorSet:         valSet,
+		roundChanges:         make(map[uint64]*qbftMsgSet),
+		prepareMessages:      make(map[uint64][]*message.SignedPreparePayload),
+		highestPreparedRound: make(map[uint64]*big.Int),
+		highestPreparedBlock: make(map[uint64]istanbul.Proposal),
+		mu:                   new(sync.Mutex),
+	}
+}
+
+type roundChangeSet struct {
+	validatorSet         istanbul.ValidatorSet
+	roundChanges         map[uint64]*qbftMsgSet
+	prepareMessages      map[uint64][]*message.SignedPreparePayload
+	highestPreparedRound map[uint64]*big.Int
+	highestPreparedBlock map[uint64]istanbul.Proposal
+	mu                   *sync.Mutex
+}
+
+func (rcs *roundChangeSet) NewRound(r *big.Int) {
+	rcs.mu.Lock()
+	defer rcs.mu.Unlock()
+	round := r.Uint64()
+	if rcs.roundChanges[round] == nil {
+		rcs.roundChanges[round] = newQBFTMsgSet(rcs.validatorSet)
+	}
+	if rcs.prepareMessages[round] == nil {
+		rcs.prepareMessages[round] = make([]*message.SignedPreparePayload, 0)
+	}
+}
+
+// Add adds the round and message into round change set
+func (rcs *roundChangeSet) Add(r *big.Int, msg message.QBFTMessage, preparedRound *big.Int, preparedBlock istanbul.Proposal, prepareMessages []*message.SignedPreparePayload, quorumSize int) error {
+	rcs.mu.Lock()
+	defer rcs.mu.Unlock()
+
+	round := r.Uint64()
+	if rcs.roundChanges[round] == nil {
+		rcs.roundChanges[round] = newQBFTMsgSet(rcs.validatorSet)
+	}
+	if err := rcs.roundChanges[round].Add(msg); err != nil {
+		return err
+	}
+
+	if preparedRound != nil && (rcs.highestPreparedRound[round] == nil || preparedRound.Cmp(rcs.highestPreparedRound[round]) > 0) {
+		roundChange := msg.(*message.RoundChange)
+		if hasMatchingRoundChangeAndPrepares(roundChange, prepareMessages, quorumSize) {
+			rcs.highestPreparedRound[round] = preparedRound
+			rcs.highestPreparedBlock[round] = preparedBlock
+			rcs.prepareMessages[round] = prepareMessages
+		}
+	}
+
+	return nil
+}
+
+// higherRoundMessages returns the number of Round Change messages received for the round greater than the given round
+// and from different validators
+func (rcs *roundChangeSet) higherRoundMessages(round *big.Int) int {
```

Comment:
I am not sure to completely understand what is this function expected to return. Could you please elaborate?

Reply:
`higherRoundMessages()` returns the number of round change messages which are higher than the given round.